src/lib/components/Swap/SwapButton.tsx

```diff
@@ -61,72 +59,73 @@ export default function SwapButton({ disabled }: SwapButtonProps) {
     useSwapApprovalOptimizedTrade(trade.trade, allowedSlippage, useIsPendingApproval) || trade.trade
 
   const approvalCurrencyAmount = useSwapCurrencyAmount(Field.INPUT)
-  const [approval, getApproval] = useSwapApproval(
-    optimizedTrade,
-    allowedSlippage,
-    useIsPendingApproval,
-    approvalCurrencyAmount
-  )
+
+  const { notApproved, pendingApproval, loadingSignature, supportsPermit, signatureData, handleApproveOrPermit } =
+    useApproveAndPermit(optimizedTrade, allowedSlippage, useIsPendingApproval, approvalCurrencyAmount)
+
   const approvalHash = usePendingApproval(
     inputCurrency?.isToken ? inputCurrency : undefined,
     useSwapRouterAddress(optimizedTrade)
   )
 
   const addTransaction = useAddTransaction()
-  const addApprovalTransaction = useCallback(() => {
-    getApproval().then((transaction) => {
-      if (transaction) {
-        addTransaction({ type: TransactionType.APPROVAL, ...transaction })
-      }
-    })
-  }, [addTransaction, getApproval])
+  const deadline = useTransactionDeadline()
```

Comment:
nit: can you move this down beneath the action props to be closer to its usage in `useSwapCallback`?